File: web/src/model/victron_controller/dashboard/Mode.ts
// @ts-nocheck
import {BaboonCodecContext, BaboonBinWriter, BinTools, BaboonBinReader, Lazy} from '../../BaboonSharedRuntime'

export enum Mode {
    Weather = "Weather",
    Forced = "Forced"
}

export const Mode_values: ReadonlyArray<Mode> = [
    Mode.Weather,
    Mode.Forced
] as const;

export function Mode_parse(s: string): Mode {
    const found = Mode_values.find(v => v === s);
    if (found === undefined) {
        throw new Error("Unknown Mode variant: " + s);
    }
    return found;
}

export class Mode_UEBACodec {
    public encode(ctx: BaboonCodecContext, value: Mode, writer: BaboonBinWriter): unknown {
        if (this !== Mode_UEBACodec.lazyInstance.value) {
          return Mode_UEBACodec.lazyInstance.value.encode(ctx, value, writer)
        }
    
        switch (value) {
            case "Weather": BinTools.writeByte(writer, 0); break;
                case "Forced": BinTools.writeByte(writer, 1); break;
            default: throw new Error("Unknown enum variant: " + value);
        }
    }
    
    public decode(ctx: BaboonCodecContext, reader: BaboonBinReader): Mode {
        if (this !== Mode_UEBACodec .lazyInstance.value) {
            return Mode_UEBACodec.lazyInstance.value.decode(ctx, reader)
        }
    
        const tag = BinTools.readByte(reader);
        switch (tag) {
            case 0: return "Weather" as Mode;
                case 1: return "Forced" as Mode;
            default: throw new Error("Unknown enum variant tag: " + tag);
        }
    }

    public static readonly BaboonDomainVersion = '0.3.0'
    public baboonDomainVersion() {
        return Mode_UEBACodec.BaboonDomainVersion
    }
    public static readonly BaboonDomainIdentifier = 'victron_controller.dashboard'
    public baboonDomainIdentifier() {
        return Mode_UEBACodec.BaboonDomainIdentifier
    }
    public static readonly BaboonTypeIdentifier = 'victron_controller.dashboard/:#Mode'
    public baboonTypeIdentifier() {
        return Mode_UEBACodec.BaboonTypeIdentifier
    }

    protected static lazyInstance = new Lazy(() => new Mode_UEBACodec())
    public static get instance(): Mode_UEBACodec {
        return Mode_UEBACodec.lazyInstance.value
    }
}

File: crates/dashboard-model/src/victron_controller/dashboard/v0_2_0/mode.rs


#[derive(Clone, Debug, PartialEq, Eq, Hash, PartialOrd, Ord, serde::Serialize, serde::Deserialize)]
pub enum Mode {
    Weather,
    Forced,
}

impl Mode {
    pub fn parse(s: &str) -> Result<Self, String> {
        match s {
            "Weather" => Ok(Mode::Weather),
            "Forced" => Ok(Mode::Forced),
            _ => Err(format!("Unknown variant: {}", s)),
        }
    }

    pub fn all() -> Vec<Mode> {
        vec![
            Mode::Weather,
            Mode::Forced,
        ]
    }
}

impl std::fmt::Display for Mode {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Mode::Weather => write!(f, "Weather"),
            Mode::Forced => write!(f, "Forced"),
        }
    }
}

impl std::str::FromStr for Mode {
    type Err = String;
    fn from_str(s: &str) -> Result<Self, Self::Err> {
        Self::parse(s)
    }
}

impl crate::baboon_runtime::BaboonBinCodecIndexed for Mode {
    fn index_elements_count(_ctx: &crate::baboon_runtime::BaboonCodecContext) -> u16 {
        0
    }
}

impl crate::baboon_runtime::BaboonBinEncode for Mode {
    fn encode_ueba(&self, _ctx: &crate::baboon_runtime::BaboonCodecContext, writer: &mut dyn std::io::Write) -> std::io::Result<()> {
        match self {
            Mode::Weather => crate::baboon_runtime::bin_tools::write_byte(writer, 0)?,
            Mode::Forced => crate::baboon_runtime::bin_tools::write_byte(writer, 1)?,
        }
        Ok(())
    }
}

impl crate::baboon_runtime::BaboonBinDecode for Mode {
    fn decode_ueba(_ctx: &crate::baboon_runtime::BaboonCodecContext, reader: &mut dyn std::io::Read) -> Result<Self, Box<dyn std::error::Error>> {
        let tag = crate::baboon_runtime::bin_tools::read_byte(reader)?;
        match tag {
            0 => Ok(Mode::Weather),
            1 => Ok(Mode::Forced),
            _ => Err(format!("Unknown enum variant tag: {}", tag).into()),
        }
    }
}